File: worker/src/xp.ts
import type { Env } from './router'
import { todayStr, nowStr, mondayOf, addDays } from './db'
import type { UserRow } from './db'
import { level_for_xp, rank_for_xp, RANKS, QUOTES_RANK, MISSIONS_DAILY, MISSIONS_WEEKLY, MISSIONS_MONTHLY } from './data'

export async function quoteOf(env: Env, occasion: string): Promise<string> {
  const rows = (await env.DB.prepare('SELECT text FROM quotes WHERE occasion = ?').bind(occasion).all()).results as { text: string }[]
  const pool = rows.length ? rows : ((await env.DB.prepare("SELECT text FROM quotes WHERE occasion = 'dia'").all()).results as { text: string }[])
  const ordinal = Math.floor(Date.now() / 86400000) + 719163
  return pool[ordinal % pool.length].text
}

export async function unlockCardsForPhase(env: Env, phaseNum: number): Promise<number> {
  const r = await env.DB.prepare('UPDATE cards SET unlocked_at = ? WHERE phase_unlock <= ? AND unlocked_at IS NULL').bind(nowStr(), phaseNum).run()
  return r.meta.changes ?? 0
}

export async function updateRank(env: Env, user: UserRow): Promise<{ slug: string; name: string } | null> {
  const [slug, name] = rank_for_xp(user.xp)
  if (user.rank_slug !== slug) {
    user.rank_slug = slug
    const msg = QUOTES_RANK[slug] ?? `Patente promovida: ${name}.`
    await env.DB.prepare('UPDATE users SET rank_slug = ? WHERE id = ?').bind(slug, user.id).run()
    await env.DB.prepare('INSERT INTO notifications (type, title, body, at, read) VALUES (?, ?, ?, ?, 0)')
      .bind('rank', `Patente promovida: ${name}`, msg, nowStr()).run()
    return { slug, name }
  }
  return null
}

export async function updateLevel(env: Env, user: UserRow): Promise<boolean> {
  const newLevel = level_for_xp(user.xp)
  if (newLevel > user.level) {
    user.level = newLevel
    await env.DB.prepare('UPDATE users SET level = ? WHERE id = ?').bind(newLevel, user.id).run()
    await env.DB.prepare('INSERT INTO notifications (type, title, body, at, read) VALUES (?, ?, ?, ?, 0)')
      .bind('level', `Nível ${newLevel} alcançado`, 'Seu nível subiu. A cidade percebeu.', nowStr()).run()
    if (newLevel >= 10) await unlockAchievement(env, user, 'nivel10')
    return true
  }
  return false
}

export async function unlockAchievement(env: Env, user: UserRow, slug: string): Promise<{ slug: string; title: string; xp: number } | null> {
  const ach = (await env.DB.prepare('SELECT * FROM achievements WHERE slug = ?').bind(slug).first()) as any
  if (ach && ach.unlocked_at === null) {
    await env.DB.prepare('UPDATE achievements SET unlocked_at = ? WHERE slug = ?').bind(nowStr(), slug).run()
    user.xp += ach.xp
    await env.DB.prepare('UPDATE users SET xp = ? WHERE id = ?').bind(user.xp, user.id).run()
    await updateLevel(env, user)
    await updateRank(env, user)
    await env.DB.prepare('INSERT INTO notifications (type, title, body, at, read) VALUES (?, ?, ?, ?, 0)')
      .bind('conquista', `Conquista: ${ach.title}`, `+${ach.xp} XP — ${ach.description}`, nowStr()).run()
    return { slug: ach.slug, title: ach.title, xp: ach.xp }
  }
  return null
}

export async function touchStreak(env: Env, user: UserRow): Promise<void> {
  const today = todayStr()
  if (user.last_study_date === today) return
  let streak: number
  if (user.last_study_date === addDays(today, -1)) streak = user.streak + 1
  else streak = 1
  const best = Math.max(user.best_streak, streak)
  user.streak = streak
  user.best_streak = best
  user.last_study_date = today
  await env.DB.prepare('UPDATE users SET streak = ?, best_streak = ?, last_study_date = ? WHERE id = ?')
    .bind(streak, best, today, user.id).run()
  if (streak === 30) await unlockAchievement(env, user, 'd30')
  if (streak === 100) await unlockAchievement(env, user, 'd100')
}

export async function missionProgress(env: Env, metric: string, amount: number): Promise<void> {
  const user = await getUserRef(env)
  if (!user) return
  const today = todayStr()
  const missions = (await env.DB.prepare('SELECT * FROM missions WHERE period_start = ? AND completed = 0').bind(today).all()).results as any[]
  for (const m of missions) {
    if (m.slug === 'm_dias25' && metric === 'minutos' && amount > 0) continue
    if (metric === 'dias' && m.slug !== 'm_dias25') continue
    let progress: number
    if (m.slug === 'm_dias25' && metric === 'dias') {
      progress = Math.min(m.progress + amount, m.target)
    } else if (['minutos', 'questoes', 'pomodoro', 'flashcards', 'redacao', 'leitura', 'revisao', 'simulado', 'precisao'].includes(metric)) {
      progress = Math.min(m.progress + amount, m.target)
    } else {
      continue
    }
    await env.DB.prepare('UPDATE missions SET progress = ? WHERE id = ?').bind(progress, m.id).run()
    if (progress >= m.target && !m.completed) {
      await env.DB.prepare('UPDATE missions SET completed = 1 WHERE id = ?').bind(m.id).run()
      user.xp += m.reward_xp
      await env.DB.prepare('UPDATE users SET xp = ? WHERE id = ?').bind(user.xp, user.id).run()
      await updateLevel(env, user)
      await updateRank(env, user)
      await env.DB.prepare('INSERT INTO notifications (type, title, body, at, read) VALUES (?, ?, ?, ?, 0)')
        .bind('missao', 'Missão cumprida', `${m.title} — +${m.reward_xp} XP`, nowStr()).run()
    }
  }
}

export async function awardXp(env: Env, user: UserRow, amount: number): Promise<number> {
  user.xp += amount
  await env.DB.prepare('UPDATE users SET xp = ? WHERE id = ?').bind(user.xp, user.id).run()
  await updateLevel(env, user)
  await updateRank(env, user)
  return user.xp
}

export async function ensureMissions(env: Env): Promise<void> {
  const today = todayStr()
  const user = await getUserRef(env)
  if (!user) return
  const daily = (await env.DB.prepare('SELECT COUNT(*) AS n FROM missions WHERE type = ? AND period_start = ?').bind('diaria', today).first()) as any
  if (!daily.n) {
    for (const [slug, title, target, , xp] of MISSIONS_DAILY) {
      await env.DB.prepare('INSERT INTO missions (type, slug, title, target, reward_xp, period_start) VALUES (?, ?, ?, ?, ?, ?)')
        .bind('diaria', slug, title, target, xp, today).run()
    }
  }
  const monday = mondayOf(today)
  const weekly = (await env.DB.prepare('SELECT COUNT(*) AS n FROM missions WHERE type = ? AND period_start = ?').bind('semanal', monday).first()) as any
  if (!weekly.n) {
    for (const [slug, title, target, , xp] of MISSIONS_WEEKLY) {
      await env.DB.prepare('INSERT INTO missions (type, slug, title, target, reward_xp, period_start) VALUES (?, ?, ?, ?, ?, ?)')
        .bind('semanal', slug, title, target, xp, monday).run()
    }
  }
  const month = today.slice(0, 8) + '01'
  const monthly = (await env.DB.prepare('SELECT COUNT(*) AS n FROM missions WHERE type = ? AND period_start = ?').bind('mensal', month).first()) as any
  if (!monthly.n) {
    for (const [slug, title, target, , xp] of MISSIONS_MONTHLY) {
      await env.DB.prepare('INSERT INTO missions (type, slug, title, target, reward_xp, period_start) VALUES (?, ?, ?, ?, ?, ?)')
        .bind('mensal', slug, title, target, xp, month).run()
    }
  }
}

async function getUserRef(env: Env): Promise<UserRow | null> {
  return (await env.DB.prepare('SELECT * FROM users ORDER BY id LIMIT 1').first()) as unknown as UserRow | null
}

export { RANKS }

File: backend/app/services/xp.py
"""Economia de XP: patentes, níveis, conquistas, missões e notificações."""
from datetime import date, datetime, timedelta
from sqlalchemy.orm import Session
from .. import curriculum
from ..models import (
    User, Achievement, Mission, Quote, Notification,
)

XP_POMODORO = 10
XP_QUESTAO = 2
XP_ACERTO = 1
XP_FLASHCARD = 1
XP_REVISAO = 5
XP_LEITURA_PAG = 1
XP_REDACAO = 50
XP_SIMULADO = 100


def quote_of(db: Session, occasion: str) -> str:
    pool = curriculum.QUOTES.get(occasion) or curriculum.QUOTES["dia"]
    day_seed = date.today().toordinal()
    return pool[day_seed % len(pool)]


def update_rank(db: Session, user: User) -> dict | None:
    slug, name, _ = curriculum.rank_for_xp(user.xp)
    promoted = None
    if user.rank_slug != slug:
        old = curriculum.RANKS[0]
        for s, n, t in curriculum.RANKS:
            if s == user.rank_slug:
                old = (s, n, t)
                break
        user.rank_slug = slug
        msg = curriculum.QUOTES["rank"].get(slug, f"Patente promovida: {name}.")
        db.add(Notification(type="rank", title=f"Patente promovida: {name}", body=msg, at=datetime.now()))
        promoted = {"slug": slug, "name": name}
    return promoted


def update_level(db: Session, user: User) -> bool:
    new_level = curriculum.level_for_xp(user.xp)
    if new_level > user.level:
        user.level = new_level
        db.add(Notification(type="level", title=f"Nível {new_level} alcançado", body="Seu nível subiu. A cidade percebeu.", at=datetime.now()))
        unlock_achievement(db, user, "nivel10") if new_level >= 10 else None
        if new_level >= 10:
            unlock_achievement(db, user, "nivel10")
        return True
    return False


def unlock_achievement(db: Session, user: User, slug: str) -> dict | None:
    ach = db.query(Achievement).filter(Achievement.slug == slug).first()
    if ach and ach.unlocked_at is None:
        ach.unlocked_at = datetime.now()
        user.xp += ach.xp
        update_level(db, user)
        update_rank(db, user)
        db.add(Notification(
            type="conquista",
            title=f"Conquista: {ach.title}",
            body=f"+{ach.xp} XP — {ach.description}",
            at=datetime.now(),
        ))
        db.commit()
        return {"slug": ach.slug, "title": ach.title, "xp": ach.xp}
    return None


def touch_streak(db: Session, user: User) -> None:
    today = date.today()
    if user.last_study_date == today:
        return
    if user.last_study_date == today - timedelta(days=1):
        user.streak += 1
    else:
        user.streak = 1
    user.last_study_date = today
    user.best_streak = max(user.best_streak, user.streak)
    if user.streak == 30:
        unlock_achievement(db, user, "d30")
    if user.streak == 100:
        unlock_achievement(db, user, "d100")


def mission_progress(db: Session, metric: str, amount: int = 1) -> None:
    """Atualiza missões por métrica e entrega XP das completas."""
    user = db.query(User).first()
    if not user:
        return
    today = date.today()
    missions = (
        db.query(Mission)
        .filter(Mission.period_start == today, Mission.completed == False)  # noqa: E712
        .all()
    )
    for m in missions:
        if m.slug in ("m_dias25",) and metric == "minutos" and amount > 0:
            continue
        if metric == "dias" and m.slug != "m_dias25":
            continue
        if m.slug == "m_dias25" and metric == "dias":
            m.progress = min(m.progress + amount, m.target)
        elif metric in ("minutos", "questoes", "pomodoro", "flashcards", "redacao", "leitura", "revisao", "simulado", "precisao"):
            m.progress = min(m.progress + amount, m.target)
        if m.progress >= m.target and not m.completed:
            m.completed = True
            user.xp += m.reward_xp
            update_level(db, user)
            update_rank(db, user)
            db.add(Notification(
                type="missao",
                title="Missão cumprida",
                body=f"{m.title} — +{m.reward_xp} XP",
                at=datetime.now(),
            ))
    db.commit()


def award_xp(db: Session, user: User, amount: int, reason: str = "") -> int:
    user.xp += amount
    update_level(db, user)
    update_rank(db, user)
    db.commit()
    return user.xp


def ensure_missions(db: Session) -> None:
    """Garante as missões de hoje/semana/mês existirem."""
    today = date.today()
    user = db.query(User).first()
    if not user:
        return
    has_daily = db.query(Mission).filter(Mission.type == "diaria", Mission.period_start == today).count()
    if not has_daily:
        day_seed = today.toordinal()
        for i, (slug, title, target, metric, xp) in enumerate(curriculum.MISSIONS_DAILY):
            db.add(Mission(type="diaria", slug=slug, title=title, target=target,
                           reward_xp=xp, period_start=today))
    monday = today - timedelta(days=today.weekday())
    has_weekly = db.query(Mission).filter(Mission.type == "semanal", Mission.period_start == monday).count()
    if not has_weekly:
        for slug, title, target, metric, xp in curriculum.MISSIONS_WEEKLY:
            db.add(Mission(type="semanal", slug=slug, title=title, target=target,
                           reward_xp=xp, period_start=monday))
    month_start = today.replace(day=1)
    has_monthly = db.query(Mission).filter(Mission.type == "mensal", Mission.period_start == month_start).count()
    if not has_monthly:
        for slug, title, target, metric, xp in curriculum.MISSIONS_MONTHLY:
            db.add(Mission(type="mensal", slug=slug, title=title, target=target,
                           reward_xp=xp, period_start=month_start))
    db.commit()
